### PR TITLE
Change to use customized Cloudwatch logger to log messages

### PR DESCRIPTION
**Pull request by hsong-rh**
Replaced `System.out.println` with cloudwatch logger.

### CODE REVIEW

File: src/main/resources/com/redhat/management/approval/MultiStageEmails.bpmn2 [hunk 1]
```diff
@@ -505,29 +515,33 @@ PostAction]]></drools:metaValue>
...
       <bpmn2:sequenceFlow id="_5CD1F091-858F-4C73-A327-4F7ADBCD6F59" drools:selectable="true" color:background-color="#000000" color:border-color="#000000" color:color="#000000" sourceRef="_F2E13387-70CE-407C-9691-747755EB76EE" targetRef="_56752DEC-FFF8-4623-AB0D-DD1D8F43D8C9"/>
       <bpmn2:sequenceFlow id="_BAC64943-752B-437C-B62E-61AD596FDB4A" drools:selectable="true" color:background-color="#000000" color:border-color="#000000" color:color="#000000" sourceRef="_E48A4947-1112-4FA4-B4A0-94A64D194D87" targetRef="_6F1ECB35-BAC7-421E-AA3C-77EEF08DAC21">
-        <bpmn2:conditionExpression xsi:type="bpmn2:tFormalExpression" id="_MTaY4mOVEeqCk5BclCu2jQ" language="http://www.java.com/java"><![CDATA[System.out.println("backOfficeStatus: "+backOfficeStatus);
-
+        <bpmn2:conditionExpression xsi:type="bpmn2:tFormalExpression" id="_MTaY4mOVEeqCk5BclCu2jQ" language="http://www.java.com/java"><![CDATA[
+Logger log = Logger.getLogger(kcontext.getVariable("insightsRequestId"));
+log.info("backOfficeStatus: "+backOfficeStatus);
```
Comment (bzwei) [inline src/main/resources/com/redhat/management/approval/MultiStageEmails.bpmn2:520]: insert space before and after `+`

File: src/main/java/com/redhat/management/approval/Logger.java [hunk 2]
```diff
@@ -61,6 +61,10 @@ public static Logger getLogger(String httpRequestId) {
     return LOGGER;
   }
 
+  public static Logger getLogger(Object httpRequestId) {
+    return httpRequestId == null ? getLogger() : getLogger(httpRequestId.toString());
```
Comment (bzwei) [inline src/main/java/com/redhat/management/approval/Logger.java:65]: `httpRequestId` can be `null` when passed to `getLogger`, no need special treatment.

**Comment by bzwei** [comment]
@hsong-rh `e.printStackTrace()` is not a good practice. I will enhance logger to take a `Throwable` as argument

File: src/main/java/com/redhat/management/approval/RequestPacket.java [hunk 3]
```diff
@@ -49,6 +49,10 @@ public String toString() {
     public String getEncodedIdentity() {
         return getHeaders().get("x-rh-identity").toString();
     }
+
+    public Object getInsightsRequestId() {
```
[the remark this answers is not in the document]
Comment (hsong-rh) [reply src/main/java/com/redhat/management/approval/RequestPacket.java:53]: The reason is key `x-rh-insights-request-id` may not be set in header. But if `getLogger()` handles `null` parameter, will add casting here.

File: src/main/java/com/redhat/management/approval/Logger.java (hunk 2 above)
Comment (hsong-rh) [reply src/main/java/com/redhat/management/approval/Logger.java:65]: 👍

**Comment by bzwei** [comment]
@hsong-rh you can add a serials new methods to the `Logger` class like the following
```
public void debug(String msg, Throwable t) {
    log("debug", throwableToString(msg, t));
}

private String throwableToString(String msg, Throwable t) {
	StringWriter buffer = new StringWriter();
	buffer.write(msg);
	buffer.write("\n");
	t.printStackTrace(new PrintWriter(buffer));
	
	return buffer.toString();
}
```

**Comment by hsong-rh** [comment]
@bzwei Added methods to accept `Throwable`

File: src/main/java/com/redhat/management/approval/Logger.java [hunk 4]
```diff
@@ -267,4 +281,13 @@ private void logToCloud(List<InputLogEvent> logs) {
       e.printStackTrace();
     }
   }
+
+  private String throwableToString(String msg, Throwable t) {
+	  StringWriter buffer = new StringWriter();
```
Comment (bzwei) [inline src/main/java/com/redhat/management/approval/Logger.java:286]: adjust the indent

File: src/main/java/com/redhat/management/approval/Logger.java [hunk 5]
```diff
@@ -98,6 +100,10 @@ public void debug(String msg) {
     log("debug", msg);
   }
 
+  public void debug(String msg, Throwable t) {
+    log("debug", throwableToString(msg, t));
+  }
+
   public void info(String msg) {
```
Comment (bzwei) [inline src/main/java/com/redhat/management/approval/Logger.java:107]: we need a `info(String, Throwable)` too

**Comment by bzwei** [comment]
@hsong-rh we need to go over all classes under `approval` to replace all `e.printStackTrace()`, for example https://github.com/RedHatInsights/approval-process/blob/master/src/main/java/com/redhat/management/approval/Request.java#L172

File: src/main/java/com/redhat/management/approval/Logger.java (hunk 5 above)
Comment (hsong-rh) [reply src/main/java/com/redhat/management/approval/Logger.java:107]: Normally `info` still need `Throwable`?

File: src/main/java/com/redhat/management/approval/Request.java [hunk 6]
```diff
@@ -183,13 +188,13 @@ private RHIdentity toRHIdentity(String encodedContext) {
         mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
         RHIdentity rhid = new RHIdentity();
 
+        Logger log = Logger.getLogger(packet.getInsightsRequestId());
```
Comment (bzwei) [inline src/main/java/com/redhat/management/approval/Request.java:191]: remove `packet`

File: src/main/java/com/redhat/management/approval/Request.java [hunk 7]
```diff
@@ -163,16 +167,17 @@ private String createEncodedIdentity(RHIdentity id) {
         obj.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
         String encoded = "";
 
+        Logger log = Logger.getLogger(packet.getInsightsRequestId());
```
Comment (bzwei) [inline src/main/java/com/redhat/management/approval/Request.java:170]: remove `packet`. consider to promote `log` to an instance member.

Comment (hsong-rh) [reply src/main/java/com/redhat/management/approval/Request.java:170]: I spent most time yesterday, tried to use instance member first. It failed due to `serializable` problem. Eventually it comes from `java.io.NotSerializableException: com.amazonaws.services.logs.AWSLogsClient`

**Pull request merged**